### PR TITLE
Clean up some pipes things

### PR DESCRIPTION
**Pull request by davidfowl**
- Use the Kestrel memory pool
- Fix bad casts

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs [hunk 1]
```diff
@@ -109,11 +109,11 @@ PipeReader IRequestBodyPipeFeature.Reader
                 if (!ReferenceEquals(_requestStreamInternal, RequestBody))
                 {
                     _requestStreamInternal = RequestBody;
-                    RequestBodyPipeReader = PipeReader.Create(RequestBody);
+                    RequestBodyPipeReader = PipeReader.Create(RequestBody, new StreamPipeReaderOptions(_context.MemoryPool, _context.MemoryPool.GetMinimumSegmentSize(), _context.MemoryPool.GetMinimumAllocSize()));
 
                     OnCompleted((self) =>
                     {
-                        ((PipeWriter)self).Complete();
+                        ((PipeReader)self).Complete();
```
Comment (rynowak) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs]: ouch

Comment (rynowak) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs]: Why isn't this defined like `OnCompleted<T>(Func<T, Task>, T)`?

**Review by jkotalik — APPROVED** [review]
Oops.

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs (hunk 1 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs]: We messed up

Comment (rynowak) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.FeatureCollection.cs]: I guess that's a good reason

**Pull request merged**